Add support for output to Apache Kafka via librdkafka

erikbeebe opened the pull request:
This pull request adds support for producing hfdl output to Apache Kafka topics (via the librdkafka driver). I've been using this in my local environment for a bit, and it works well (and makes the data easy to work with using third party stream processors, etc).

(This PR builds on the previous one to add a couple bug fixes, a few logging fixes, configurables, etc).

I've tested this build on Linux (Ubuntu 22.04.5 LTS) with the Ubuntu-packaged librdkafka 1.8.0-1build1, tested against brokers with and without authentication.

Thanks for building dumphfdl - it's a lot of fun to use!

szpajder commented:
```
In file included from /home/szpajder/src/dumphfdl/dumphfdl/src/output-rdkafka.c:10:
/home/szpajder/src/dumphfdl/dumphfdl/src/output-rdkafka.c: In function ‘rdkafka_conf_set’:
/home/szpajder/src/dumphfdl/dumphfdl/src/output-rdkafka.c:80:18: error: ‘null’ undeclared (first use in this function)
   80 |   ASSERT(conf != null);
make[2]: *** [src/CMakeFiles/dumphfdl_base.dir/build.make:629: src/CMakeFiles/dumphfdl_base.dir/output-rdkafka.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:117: src/CMakeFiles/dumphfdl_base.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

szpajder commented on src/output-rdkafka.c:
```diff
@@ -0,0 +1,230 @@
+        RD_KAFKA_V_MSGFLAGS(RD_KAFKA_MSG_F_COPY),
+        RD_KAFKA_V_OPAQUE(&delivery_counter),
+        RD_KAFKA_V_END);
+    if (err) {
```
How about `err != RD_KAFKA_RESP_ERR_NO_ERROR`?

erikbeebe replied:
Good point, I based mine off of the Confluent producer example here:

https://github.com/confluentinc/librdkafka/blob/34fb3e3071509293552680bdb3e256ac70608f87/examples/producer.c#L172

but RD_KAFKA_RESP_ERR_NO_ERROR seems clearer - updated

szpajder commented on src/output-rdkafka.c:
```diff
@@ -0,0 +1,230 @@
+        RD_KAFKA_V_OPAQUE(&delivery_counter),
+        RD_KAFKA_V_END);
+    if (err) {
+        fprintf(stderr, "output_rdkafka: Produce failed: %s\n",
```
`"output_rdkafka(%s): ...", self->brokers` (in case there is more than one kafka output, the user should now which output the error message applies to).

erikbeebe replied:
Good point, certainly possible - updated to make this clearer

erikbeebe commented on src/output-rdkafka.c:
```diff
@@ -0,0 +1,230 @@
+      cfg->acks = "all";
+    }
+
+    // Output a log message to stderr for each message produced to Kafka
```
Fair point.  I run five of these in tmux (one for each SDR), so I like being able to switch through them quickly to see which one is producing messages quickly while I'm debugging.. debug_print() seems ideal for this, I'll update to use this instead

erikbeebe commented on src/output-rdkafka.c:
```diff
@@ -0,0 +1,230 @@
+    if(format == OFMT_TEXT || format == OFMT_JSON || format == OFMT_BASESTATION) {
+        out_rdkafka_produce_text(self, metadata, msg);
+    }
+    return 0;
```
Ah ok, I didn't realize it would retry - I took a look at the zmq code as an example, and updated this accordingly

erikbeebe commented on src/output-rdkafka.c:
```diff
@@ -0,0 +1,230 @@
+/* SPDX-License-Identifier: GPL-3.0-or-later */
+#include <stdio.h>                      // fprintf
+#include <string.h>                     // strdup, strerror
+#include <errno.h>                      // errno
```
Removed

erikbeebe commented on src/output-rdkafka.c:
```diff
@@ -0,0 +1,230 @@
+#include <errno.h>                      // errno
+#include <librdkafka/rdkafka.h>
+#include "output-common.h"              // output_descriptor_t, output_qentry_t, output_queue_drain
+#include "config.h"                     // Uhh
```
Removed

erikbeebe commented on src/output-rdkafka.c:
```diff
@@ -0,0 +1,230 @@
+    char *security_protocol;
+    char *acks;
+    bool verbose_log;
+    void *rdkafka_ctx;
```
Removed

erikbeebe commented on src/output-rdkafka.c:
```diff
@@ -0,0 +1,230 @@
+    char *acks;
+    bool verbose_log;
+    void *rdkafka_ctx;
+    void *rdkafka_sock;
```
Removed

erikbeebe commented on src/output-rdkafka.c:
```diff
@@ -0,0 +1,230 @@
+    rd_kafka_resp_err_t err;
+
+    if (self->verbose_log) {
+      fprintf(stderr, "output_rdkafka: producing message to kafka topic: %s\n", self->topic);
```
fair enough - removed

erikbeebe replied:
(Just using a file:path=- output to write the messages to stdout achieves this anyway I discovered, so probably a better option overall)

szpajder commented:
So now I'm trying to induce some error conditions to see how the program would behave on a non-happy path. It seems that the current error handling doesn't do much, because librdkafka hides most of the errors from the caller and either prints its own error messages to stderr (which is acceptable) or stays completely silent (which is no good, because it hides problems from the user and makes troubleshooting more difficult).

Example 1: attempt to connect to a port where broker is not listening:

```
output_rdkafka(localhost:9091): creating producer...
output_rdkafka(localhost:9091): connection established
%3|1746478333.503|FAIL|rdkafka#producer-1| [thrd:localhost:9091/bootstrap]: localhost:9091/bootstrap: Connect to ipv6#[::1]:9091 failed: Connection refused (after 0ms in state CONNECT)
%3|1746478333.503|ERROR|rdkafka#producer-1| [thrd:localhost:9091/bootstrap]: 1/1 brokers are down
%3|1746478333.627|ERROR|rdkafka#producer-1| [thrd:app]: rdkafka#producer-1: localhost:9091/bootstrap: Connect to ipv6#[::1]:9091 failed: Connection refused (after 0ms in state CONNECT)
%3|1746478334.504|FAIL|rdkafka#producer-1| [thrd:localhost:9091/bootstrap]: localhost:9091/bootstrap: Connect to ipv4#127.0.0.1:9091 failed: Connection refused (after 0ms in state CONNECT)
%3|1746478334.536|ERROR|rdkafka#producer-1| [thrd:app]: rdkafka#producer-1: localhost:9091/bootstrap: Connect to ipv4#127.0.0.1:9091 failed: Connection refused (after 0ms in state CONNECT)
%3|1746478335.503|FAIL|rdkafka#producer-1| [thrd:localhost:9091/bootstrap]: localhost:9091/bootstrap: Connect to ipv4#127.0.0.1:9091 failed: Connection refused (after 0ms in state CONNECT, 1 identical error(s) suppressed)
%3|1746478335.532|ERROR|rdkafka#producer-1| [thrd:app]: rdkafka#producer-1: localhost:9091/bootstrap: Connect to ipv4#127.0.0.1:9091 failed: Connection refused (after 0ms in state CONNECT, 1 identical error(s) suppressed)
Waiting for all threads to finish
Shutting down decoder thread
output_rdkafka(localhost:9091): shutting down
%4|1746478336.499|TERMINATE|rdkafka#producer-1| [thrd:app]: Producer terminating with 23 messages (8280 bytes) still in queue or transit: use flush() to wait for outstanding message delivery
```

"connection established", but not really...

The final message is also interesting - it suggests that librdkafka queues messages on its own. So maybe returning an error and keeping the message in dumphfdl's queue for redelivery might not be desirable and could cause duplicates, at least in theory.

Example 2: attempt to write to a topic which does not exist on the cluster (with `auto.create.topics.enable` set to `false`) yields absolutely no error messages when the program is running. Only on quit there is this message about producer terminating with messages still enqueued.

```
dumphfdl 1.6.1-d6834f5 (libacars 2.2.0)
output_rdkafka: Configuring broker: localhost:9092
Aircraft cache TTL set to 3600 seconds
output_rdkafka(localhost:9092): creating producer...
output_rdkafka(localhost:9092): connection established
Waiting for all threads to finish
Shutting down decoder thread
output_rdkafka(localhost:9092): shutting down
%4|1746478393.234|TERMINATE|rdkafka#producer-1| [thrd:app]: Producer terminating with 23 messages (8280 bytes) still in queue or transit: use flush() to wait for outstanding message delivery
```

Given that this is a permanent error, it should be reported somehow. The [manual](https://github.com/confluentinc/librdkafka/wiki/Error-handling) says that `error_cb()` callback shall be defined for this purpose.

erikbeebe commented:
It's definitely optimistic about eventual success in some places where it won't be.  Let me dig in a bit later and I'll try to make some of that more robust.

erikbeebe commented:
If you get a few minutes, please give those a try and see how they work.  I've done some testing against my broker here (setting invalid auth, etc), or setting invalid broker or port, and the initial synchronous check seems to do a good job of catching configuration problems, and the error_cb() should make transient errors visible.  Thanks for the feedback.

erikbeebe commented:
Quick update - I've been running my five instances of dumphfdl for a few days with these modifications, and it's worked fine, handled reconnects correctly, etc.  As an experiment, I updated my Ubuntu 22.04 LTS development environment from librdkafka 1.8.0 (the one currently packaged in the apt repository) to librdkafka master (~2.10), and it compiles and works fine with no modifications as well.

szpajder commented:
Thanks for the updates.

In no circumstances should output handlers call `exit()`. If you have, say, twenty different outputs, why would you want the program to exit only because one of them has failed?

Return -1 instead. The output thread will then disable the output and terminate itself and the program will happily continue on all remaining outputs. Here is an example of a failed `file` output initialization:

```
Could not open output file /cantwritehere.log: Permission denied
output_file: could not write to '/cantwritehere.log', deactivating output
```

The only place where you can't do it is the error callback - but it's not a big deal, essentially every callbacked fatal rdkafka error will be treated as transient. This is not an issue at all, provided that the user gets informed properly on what went wrong.

As far as connection failures go - they may happen for various reasons, not necessarily due to a user error. Let's say dumphfdl is started on system boot and the network connection needs some time to fully initialize (ie. because it's a WLAN or cellular network or whatever). If rdkafka treats connection failure as a transient error and attempts to reconnect by itself, then this error shall not be turned into a fatal one. Otherwise the program may exit before the network connection is set up. I liked the previous solution better (ie. let rdkafka print its "Connection refused" message and continue). The only problem was the "connection established" message which was confusing.

I still don't get any error message when attempting to produce to a nonexistent topic. I didn't have time yet to dig into librdkafka innards to find out how it would handle this, but in theory it should be able to. This condition definitely gets reported by the broker - in the Metadata response message:

```
    Topic Metadata
        Topic (hfdl2)
            Error: Unknown Topic or Partition (3)
            Topic Name: hfdl2
            Topic ID: 00000000-0000-0000-0000-000000000000
            Is Internal: False
            Topic Authorized Operations: 0x80000000
            Tagged fields
```

erikbeebe commented:
For non-existent topics - 

Today, if we produce to a topic that does not exist, in a cluster with ```auto.create.topics.enable``` set to false, we log:

```output_rdkafka(localhost:42129): Produce message failed: Local: Unknown topic [0/1854]```

If we produce to a topic which does not exist, in a cluster with ```auto.create.topics.enable``` set to true, the topic will be created (with the default number of partitions configured for the cluster).

We can detect these and log something a little more obvious if we implement it in a delivery report callback (but can't otherwise solve the problem of course).

erikbeebe commented:
I checked in a few updates that should address these items.  In particular:

- Updated all error handling that previously called exit(), to return to dumphfdl for retry
- Implemented the message delivery report callback to make errors (particularly topic) clearer

erikbeebe commented:
Yeah, most of the value from the delivery callback is actually when it *does* work (since it enables us to print out useful debug information like the partition and offset that message was assigned, making it easier to debug what the partitioner is doing if we care about ordering, etc) - but since we're not adding any verbose logging, and don't set a message key right now, it's pretty optional.
